src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs

```
@@ -120,18 +137,28 @@ public async Task StopAsync(CancellationToken cancelToken = default)
             }
         }
 
-        private int GetRandomPort()
+        private int GetLikelyAvailablePort()
         {
-            // Get a random port that has not already been attempted so we don't waste time trying
-            // to listen to a port we know is not free.
+            // Get an available port for use in the gRPC server. Try 4001 first, then select a random open port
+            // in the 30000-31000 range.
+            IPGlobalProperties ipGlobalProperties = IPGlobalProperties.GetIPGlobalProperties();
+            TcpConnectionInformation[] tcpConnInfoArray = ipGlobalProperties.GetActiveTcpConnections();
+            if (!tcpConnInfoArray.Where(x => x.LocalEndPoint.Port == DefaultPort).Any())
+            {
+                return DefaultPort;
+            }
+
             int randomPort;
-            do
+            for (int i = 0; i < 10; i++)
             {
                 randomPort = this.portGenerator.Next(MinPort, MaxPort);
+                if (!tcpConnInfoArray.Where(x => x.LocalEndPoint.Port == randomPort).Any())
+                {
+                    return randomPort;
+                }
             }
-            while (this.attemptedPorts.Contains(randomPort));
 
-            return randomPort;
+            return 0;
```

Probably need to either throw here, or guard against this in the calling method. Any preference?

---

Yeah, we should throw so that we don't get into some unknown state. But I wonder if we should have a larger number of attempts. Inevitably, throwing will restart the host, which triggers more retries anyways.

---

Bumped attempts to 50 and added an exception if no port is found in 50 attempts. Might be able to remove all of this if we abandon the 30000-31000 requirement as detailed above